### PR TITLE
add Kinesis Transport and support config Kinesis in Spark integration

### PR DESCRIPTION
**Pull request by yogyang**
Signed-off-by: yogyang <yogayang.h@gmail.com>

### Problem

Closes: https://github.com/OpenLineage/OpenLineage/issues/1199

### Solution

add new dependency in client/java
```
https://github.com/OpenLineage/OpenLineage/issues/1199
```

Test the Spark-Integration to Kinesis in Databricks  + AWS Kinesis

### Checklist

- [x] You've [signed-off](https://github.com/OpenLineage/OpenLineage/blob/main/why-the-dco.md) your work
- [x] Your pull request title follows our [guidelines](https://github.com/OpenLineage/OpenLineage/blob/main/CONTRIBUTING.md#creating-pull-requests)
- [ ] Your changes are accompanied by tests (_if relevant_)
- [ ] Your change contains a [small diff](https://kurtisnusbaum.medium.com/stacked-diffs-keeping-phabricator-diffs-small-d9964f4dcfa6) and is self-contained
- [ ] You've updated any relevant documentation (_if relevant_)
- [ ] You've updated the [`CHANGELOG.md`](https://github.com/OpenLineage/OpenLineage/blob/main/CHANGELOG.md) with details about your change under the "Unreleased" section (_if relevant, depending on the change, this may not be necessary_)
- [ ] You've versioned the core OpenLineage model or facets according to [SchemaVer](https://docs.snowplowanalytics.com/docs/pipeline-components-and-applications/iglu/common-architecture/schemaver) (_if relevant_)
- [x] You've added a [header](https://github.com/OpenLineage/OpenLineage/tree/main/.github/header_templates.md) to source files (_if relevant_)

### CODE REVIEW

File: integration/spark/app/src/main/java/io/openlineage/spark/agent/OpenLineageSparkListener.java [hunk 1]
```diff
@@ -297,10 +302,53 @@ private static boolean checkIfDisabled() {
...
     Optional<String> url = findSparkConfigKey(conf, SPARK_CONF_URL_KEY);
+    Optional<String> transportType = findSparkConfigKey(conf, SPARK_CONF_TRANSPORT_TYPE);
     if (url.isPresent()) {
       return ArgumentParser.parse(url.get());
+    } else if (transportType.isPresent()) {
+      // go and check the specific transport type setting
+      String mode = transportType.get();
+      Optional<TransportConfig> transportConfig = Optional.empty();
+      switch (mode) {
+        case "Kinesis":
```
Comment (pawel-big-lebowski) [inline integration/spark/app/src/main/java/io/openlineage/spark/agent/OpenLineageSparkListener.java:329]: Would it make sense to do lower-case when comparing ?

**Review by pawel-big-lebowski — COMMENTED** [review]
Wow, it is great to see such a PR with that sweet feature 🚀 👍 

Please provide some documentation on this
 * here -> https://github.com/OpenLineage/OpenLineage/blob/main/integration/spark/README.md?plain=1#L105
 * and here -> https://github.com/OpenLineage/docs/blob/main/docs/integrations/spark/spark.md?plain=1#L87

Sorry for having two copies of the same content. 

Adding new entry to CHANGELOG file would be great.

File: client/java/src/test/java/io/openlineage/client/transports/KinesisTransportTest.java [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+  @Test
+  void clientEmitsKinesisTransport() throws IOException {
+    KinesisProducer producer = mock(KinesisProducer.class);
+    KinesisConfig config = new KinesisConfig();
+
+    Properties properties = new Properties();
+    properties.setProperty("MinConnections", "1");
+
+    config.setRegion("us-west-2");
+    config.setRoleArn("test-role");
```
Comment (mobuchowski) [inline client/java/src/test/java/io/openlineage/client/transports/KinesisTransportTest.java:33]: This should be optional - it's failing test https://app.circleci.com/pipelines/github/OpenLineage/OpenLineage/4949/workflows/081d660c-e5d6-432c-9eda-3d33087a8b23/jobs/62505?invite=true#step-103-68

File: client/java/src/main/java/io/openlineage/client/transports/KinesisTransport.java [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+import java.nio.ByteBuffer;
+import java.util.Optional;
+import java.util.concurrent.Executor;
+import java.util.concurrent.Executors;
+import lombok.NonNull;
+import lombok.extern.slf4j.Slf4j;
+
+@Slf4j
+public class KinesisTransport extends Transport {
+  private final String streaName;
```
Comment (mobuchowski) [inline client/java/src/main/java/io/openlineage/client/transports/KinesisTransport.java:22]:
```suggestion
  private final String streamName;
```

File: client/java/src/main/java/io/openlineage/client/transports/KinesisTransport.java [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+
+  private final KinesisProducer producer;
+
+  private final Executor listeningExecutor;
+
+  public KinesisTransport(
+      @NonNull final KinesisProducer kinesisProducer, @NonNull final KinesisConfig kinesisConfig) {
+
+    super(Type.KINESIS);
+    this.streaName = kinesisConfig.getStreamName();
```
Comment (mobuchowski) [inline client/java/src/main/java/io/openlineage/client/transports/KinesisTransport.java:34]:
```suggestion
    this.streamName = kinesisConfig.getStreamName();
```

File: client/java/src/main/java/io/openlineage/client/transports/KinesisTransport.java [hunk 5]
```diff
@@ -0,0 +1,83 @@
...
+    this.region = kinesisConfig.getRegion();
+    this.roleArn = kinesisConfig.getRoleArn();
+    this.producer = kinesisProducer;
+    this.listeningExecutor = Executors.newSingleThreadExecutor();
+  }
+
+  public KinesisTransport(@NonNull final KinesisConfig kinesisConfig) {
+    super(Type.KINESIS);
+    this.streaName = kinesisConfig.getStreamName();
```
Comment (mobuchowski) [inline client/java/src/main/java/io/openlineage/client/transports/KinesisTransport.java:43]:
```suggestion
    this.streamName = kinesisConfig.getStreamName();
```

File: client/java/src/main/java/io/openlineage/client/transports/KinesisTransport.java [hunk 6]
```diff
@@ -0,0 +1,83 @@
...
+  }
+
+  @Override
+  public void emit(@NonNull OpenLineage.RunEvent runEvent) {
+    final String eventAsJson = OpenLineageClientUtils.toJson(runEvent);
+    log.debug("Received lineage event: {}", eventAsJson);
+    ListenableFuture<UserRecordResult> future =
+        this.producer.addUserRecord(
+            new UserRecord(
+                streaName,
```
Comment (mobuchowski) [inline client/java/src/main/java/io/openlineage/client/transports/KinesisTransport.java:64]:
```suggestion
                streamName,
```

File: client/java/src/main/java/io/openlineage/client/transports/Transport.java [hunk 7]
```diff
@@ -13,7 +13,8 @@ enum Type {
     CONSOLE,
     HTTP,
     KAFKA,
-    NOOP
+    NOOP,
+    KINESIS
```
Comment (mobuchowski) [inline client/java/src/main/java/io/openlineage/client/transports/Transport.java:17]:
```suggestion
    KINESIS,
    NOOP
```

**Review by mobuchowski — CHANGES_REQUESTED** [review]
Few small changes, looks good besides that 😉 

@yogyang have you tested it manually with Kinesis? We don't have AWS account to integration test those, at least for now.

File: integration/spark/app/src/main/java/io/openlineage/spark/agent/OpenLineageSparkListener.java (hunk 1 above)
Comment (yogyang) [reply integration/spark/app/src/main/java/io/openlineage/spark/agent/OpenLineageSparkListener.java:329]: sure

**Comment by yogyang** [comment]
> Few small changes, looks good besides that 😉
> 
> @yogyang have you tested it manually with Kinesis? We don't have AWS account to integration test those, at least for now.

Yeah, I build the spark integration jar and test in my env with Databricks with Kinesis, it send successfully.

**Comment by mobuchowski** [comment]
@yogyang thanks for contribution and testing that. 

Right now, `spotless` is failing so please commit changes made by `./gradlew spotlessApply`. Unfortunately `spotless` have different behavior depending on Java version, so you might already done that and got wrong results 😞 

Spark integration depends on Java 8. 

Also, please follow https://github.com/OpenLineage/OpenLineage/pull/1200/checks?check_run_id=9266333659 to have correct DCO sign of commits - first two commits are correctly signed, but the last one isn't.

**Comment by yogyang** [comment]
Hi @mobuchowski  @pawel-big-lebowski , Thanks for feedback. 
I think I've made enough changes, but the CI `ci/circleci: integration-test-integration-spark-3.2.2` failed on Timeout, It seems I can not rerun it.

**Review by mobuchowski — APPROVED** [review]

**Comment by mobuchowski** [comment]
@yogyang I've retried and it passed. The PR looks good to me, thanks for contributing 👍

**Pull request merged**